src/metabase/models/card.clj

```diff
@@ -551,12 +551,70 @@
       (pre-update-check-sandbox-constraints changes)
       (assert-valid-type (merge old-card-info changes)))))
 
+(defn- derive-ident [prefix entity_id stage-number tail]
+  ;; XXX: Probably bring this back, though it's likely to break a load of tests because I didn't specify the joins
+  ;; and especially breakouts as aggressively as I might have done.
+  ;; Alternatively, add a dynamic var or something for ad-hoc queries, since this might happen for them?
+  (when-not entity_id
+    (throw (ex-info "derive-ident with blank entity_id" {:prefix prefix :stage-number stage-number :tail tail})))
```

Comment:
This is an important check, but I'm not sure if it'll break lots of tests or not.